### PR TITLE
Added CySignals

### PR DESCRIPTION
**Pull request by ShikharJ**

### CODE REVIEW

**Comment by isuruf** [comment]
You should use cysignals as an external library.

**Comment by ShikharJ** [comment]
@isuruf Should it be an optional addition?

**Comment by isuruf** [comment]
Yes.

**Comment by ShikharJ** [comment]
@isuruf Please take a look and tell me if this is fine. Also, for using `Cysignals`, I'll have to find discrete `for` and `while` loops and use `sig_check()` or `sig_on()`/`sig_off()` calls right?

**Comment by ShikharJ** [comment]
ping @isuruf

**Comment by ShikharJ** [comment]
Ping @isuruf.

**Comment by isuruf** [comment]
this should be an option to setup.py, but we can work on that later when we have some code that use cysignals.

**Comment by ShikharJ** [comment]
Can you please take a look @isuruf ?

File: symengine/lib/symengine_wrapper.pyx [hunk 1]
```diff
@@ -16,6 +16,10 @@ import warnings
 
 include "config.pxi"
 
+from setup import use_cysignals
+if use_cysignals:
+    include "cysignals/signals.pxi"
```
Comment (isuruf) [inline symengine/lib/symengine_wrapper.pyx]: Does this work?

Comment (ShikharJ) [inline symengine/lib/symengine_wrapper.pyx]: I actually wanted to know whether there is a workaround for this?
I'm not sure this would work.

**Comment by ShikharJ** [comment]
@isuruf Is there a possible workaround for this?

File: symengine/lib/symengine_wrapper.pyx (hunk 1 above)
Comment (isuruf) [inline symengine/lib/symengine_wrapper.pyx]: You need to use compile time constants like what we do with constant like `HAVE_SYMENGINE_MPFR`

**Comment by ShikharJ** [comment]
@isuruf The `Cysignals` [docs](http://cysignals.readthedocs.io/en/latest/) suggest adding `include_path=sys.path` to the `cythonize()` call ([like this](https://github.com/sagemath/cysignals/blob/master/example/setup.py#L8)). However, I couldn't find a `cythonize()` call in the repository. How can this be overcome?

**Comment by ShikharJ** [comment]
Ping @isuruf.

**Comment by ShikharJ** [comment]
Ping @isuruf

**Comment by isuruf** [comment]
`sig_on` and `sig_off` should be added before and after every call to symengine as symengine methods can take a long time

**Comment by isuruf** [comment]
Also, is using cysignals safe with `PyNumber` and `PyFunction` ? (They call the Python C API)

**Comment by ShikharJ** [comment]
@isuruf The docs mention that `sig_on()` and `sig_off()` should only be used with pure C or pure Cython code. `sig_check()` is relatively much safer, as `sig_on`/`sig_off` do not perform memory cleanup on being executed. As a safe route the docs also say: 
```
When in doubt, try to use sig_check() instead.
```
So, I think they should be safe to use with `PyNumber` and `PyFunction`. 
I'll make the changes for `SymEngine` calls

**Comment by ShikharJ** [comment]
@isuruf Can you take a look here? Also, what is to be done regarding [this](https://github.com/symengine/symengine.py/pull/127#issuecomment-286037378) ?

**Comment by ShikharJ** [comment]
Ping @isuruf.

**Comment by isuruf** [comment]
> So, I think they should be safe to use with PyNumber and PyFunction.

Why?

After reading the docs you mentioned, I think it's not really safe to use CySignals because of the Python/C API calls we make

**Comment by ShikharJ** [comment]
@isuruf But would those calls be (majorly) different in mechanism to our calls to the external libraries?

**Comment by isuruf** [comment]
For example, `diff` function will call the C++ library method, but it will callback to Python if a `PyFunction` or `PyNumber` is found in the expression tree.

**Comment by isuruf** [comment]
Sorry, wrong button

**Comment by ShikharJ** [comment]
Ah! I completely overlooked that. I'll update this. Can you also tell me what to do about [this](https://github.com/symengine/symengine.py/pull/127#issuecomment-286037378) ?

**Comment by isuruf** [comment]
That's fine. We are not using `cythonize`.

**Pull request closed without merging**